### DISCUSSION TITLE
要約と疑問点

### DISCUSSION THREAD

**Discussion by Kaito1290** [Progaku-copy/isucon-knowledge#2] [ISUCON本第１章 輪読会]
### 高速なWebサービスとは?
- レイテンシが低く、スループットが高い
⇒リクエストに対する待ち時間が短く、同時にたくさんのリクエストを処理できる。キャパシティが十分であれば上記を満たすことができる。

- 高速であればコスト対性能効率が良い
⇒高速に処理できるということはサーバーのリソースを占有する時間が短いので、同じリソースでも単位時間あたりにたくさんの処理を実行できる。高速なシステムはムダがないということ。

### 必要十分なキャパシティを用意するには
- 必要十分なキャパシティ
⇒需要に対してキャパシティが不足せず、過剰でない状態。

- 需要側の調整
⇒処理要求を順番待ちさせる、処理要求の受付を単位時間あたり一定数に絞る、処理要求の発生にランダムな待ち時間を発生させる手法がある。

- 供給側の調整
⇒垂直スケーリング、水平スケーリングを事前に行うプロアクティブな対応と、オートスケーリングでシステムリソース需要をもとに調整するリアクティブな対応がある。

### パフォーマンスチューニングのきほん
- パフォーマンスチューニングの流れ
⇒「負荷試験」、「結果の確認」、「システムの改善」を繰り返すことが基本。

- 負荷試験
⇒ Webサービスを構成する要素それぞれのシステムリソース利用状況、レイテンシ、スループットなどを計測する。
　1つずつ対策を適用して試験することで有意なデータを取ることができる。

- 結果の確認
⇒結果を比較するときは前提条件を揃える。結果に含まれている誤差やノイズに注意する。

- システムの改善
⇒ボトルネックを特定して対処することが大切。ボトルネックとなっている課題は「解決 / 回避 / 緩和」のどれかで対処する。
---
### 1章で出てきた用語
- レイテンシ 
クライアントがリクエストしてから、受信完了するまでの待機時間。
    
- スループット
リクエストの同時並行処理性能。単位は単位時間あたりのリクエスト処理数rps(request per second)
    
- パフォーマンスチューニング
高速なWebサービスを実現するための取り組み
    
- 垂直スケーリング
サーバーの性能を向上させてシステムリソースの総量を変える（スケールアップorスケールダウン）手法
    
- 水平スケーリング
サーバーの数を変える（スケールアウト、スケールイン）手法
    
- オートスケーリング
インフラのスケール変更を自動で行う機能

**Comment by Kaito1290** [comment]
### 疑問点
ISCONではインフラが決まっているということですが、
必要十分なキャパシティを用意するという観点から高速化を考えるとキャパシティの供給側の調整はできなくて、需要側の調整をするという認識であってますか？

**Reply by Kiyo510** [reply]
↑その通り。

### ISUCONのリソースは固定
- コンテナ数は限られていて、EC2が3台あたえられる。c5.large。
- 最初はAP・DB・WEBサーバーはひとつに同居している。
- リソースはそもそもつくるところから
  - CloudFormationをこっちで用意しておく

**Comment by kakeru-one** [comment]
https://github.com/isucon/isucon13
